### PR TITLE
Implement dqn

### LINKED ISSUE #5: Implement DQN

### PR DESCRIPTION
**Pull request by ngorke**
closes #5

### CODE REVIEW

File: agent.py [hunk 1]
```diff
@@ -121,12 +124,10 @@ def get_optimal_action(self, observation) -> int:
     
     def update(self):
         action = self.__get_random_action()
+        self.__step += 1
```
Comment (magsteen) [inline agent.py]: remove

File: agent.py [hunk 2]
```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
...
+        self.epsilon_decay = 500
+
+        self.policy_net = self.nn
+        self.target_net = deepcopy(self.nn)
+
+        self.replay_memory = ReplayMemory(self.memory_size, self.channels, self.grid_size)
+        self.tensorized_obs = self.__tensorize_observation(ctx.env.reset())
+
+        self.policy_net.train(True)
+        self.target_net.train(False)
```
Comment (chrisryda) [inline agent.py]: Add name-mangling to all vars

File: agent.py [hunk 3]
```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
...
+            s[i] = torch.where(t == i + 1, t // (i + 1), 0)
+        return s
+
     @property
     def info(self) -> dict:
         return {
             **super().info,
-            "discretizer": self.__discretizer.info,
+            "nn": self.nn.info,
         }
```
Comment (chrisryda) [inline agent.py:288]: Check through DQNAgent that all vars/methods that should be _ or __, are :eyes:

File: graphing.py [hunk 4]
```diff
@@ -28,7 +28,7 @@ def _get_chunk_size(self):
     
     def _reduce_to_avg(self, list: list, chunk_size: int):
         """Divides a long list of values into chunks and finds the average value of each chunk."""
-        chunks = [list[i:i + chunk_size] for i in range(0, len(list), chunk_size)]
+        chunks = [list[i:(i + chunk_size)] for i in range(0, len(list), chunk_size)]
         return [sum(chunks[i]) / len(chunks[i]) for i in range(len(chunks))]
```
Comment (chrisryda) [inline graphing.py]: _reduce_to_best has for i in range(len(chunks) - 1) - this should be the same for both methods - pick one!

File: main.py [hunk 5]
```diff
@@ -8,6 +9,7 @@
 from aioconsole import ainput, aprint
 from dataclasses import dataclass
 from discretizer import DiscretizerType, FullDiscretizer, QuadDiscretizer, AngularDiscretizer
+from dqn import ModelType, LinearDQN, ConvolutionalDQN
 import sys
 from graphing import StatsType
```
Comment (chrisryda) [inline main.py]: Switch places of these two :see_no_evil:

File: agent.py [hunk 6]
```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
...
+            q_arrays = self.target_net(new_states)
+        q_max, _ = torch.max(q_arrays, axis=1, keepdim=True)
+        q_values = torch.multiply(q_max, self.gamma) + rewards.view(-1, 1)
+
+        self.optimizer.zero_grad()
+        self.policy_net.loss(states, actions, q_values).backward()
+
+        for param in self.policy_net.parameters():
+            param.grad.data.clamp(-1, 1)
+        self.optimizer.step()
```
Comment (ngorke) [inline agent.py]: Refactor into own method

File: agent.py [hunk 7]
```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
...
+
+        self.steps += 1
+        if self.steps % self.T == 0:
+            self.__copy_q_to_target()
```
Comment (ngorke) [inline agent.py]: Refactor into own method

File: replay_memory.py [hunk 8]
```diff
@@ -1,17 +1,41 @@
...
+    def push(self, state, new_state, action, reward):
+        """Save a transition, and if full, overwrites the leftmost transition"""
+        self.__len = min(self.__len + 1, self.__cap)
+        self.__states[self.__write_index] = state
+        self.__new_states[self.__write_index] = new_state
+        self.__actions[self.__write_index] = action
+        self.__rewards[self.__write_index] = reward
+        self.__write_index = (self.__write_index + 1) % self.__cap
+    
+    def random_indices(self, batch_size):
```
Comment (chrisryda) [inline replay_memory.py]: Rename to get_random_indices() ?

File: agent.py [hunk 9]
```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
...
+                    + (self.epsilon_start - self.epsilon_end) \
+                    * np.exp(-1 * self.steps / self.epsilon_decay)
+                    
+        if np.random.random() < epsilon:
+            return self.__get_random_action()
+        else:
+            action_q_vals = None
+            with torch.no_grad():
+                action_q_vals = self.policy_net(self.tensorized_obs.view(1, self.channels, self.grid_size, self.grid_size))
+            return torch.argmax(action_q_vals).item()
```
Comment (ngorke) [inline agent.py]: Else is not needed

File: agent.py [hunk 10]
```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
...
+        
+        return observation, reward, terminated, truncated, info
+
+    def __experience_initial(self):
+        for _ in range(20_000):
+            self.__experience_replay(explore_only=True)
+        self.tensorized_obs = self.__tensorize_observation(self.env.reset())
+
+    def __copy_q_to_target(self):
+        self.target_net.load_state_dict(deepcopy(self.policy_net.state_dict()))
```
Comment (ngorke) [inline agent.py]: Do we really need to deepcopy? I saw torch did not use it

File: transition.py [hunk 11]
```diff
@@ -4,3 +4,9 @@ def __init__(self, state, new_state, action, reward):
         self.new_state = new_state
         self.action = action
         self.reward = reward
+
+    def __str__(self):
+        return f'state: {self.state}, new_state: {self.new_state}, action: {self.action}, reward: {self.reward}\n'
```
Comment (chrisryda) [inline transition.py]: Use double-quotes!

**Review by chrisryda — CHANGES_REQUESTED** [review]
Amazing, truly amazing work - see comments<3

File: agent.py [hunk 12]
```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
...
+        self.optimizer = torch.optim.AdamW(self.policy_net.parameters(), lr=self.alpha)
+
+    def get_optimal_action(self, observation) -> int:
+        observation = self.__tensorize_observation(observation)
+        with torch.no_grad():
+            action_q_vals = self.policy_net(observation.view(1, self.channels, self.grid_size, self.grid_size)) 
+        return torch.argmax(action_q_vals).item()
+
+    def initialize(self):
+        self.__experience_initial()
```
Comment (magsteen) [inline agent.py]: This wrapper is unnecessary? The "__experience_initial" function's functionality should be moved here

File: agent.py [hunk 13]
```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
...
+        self.tensorized_obs = new_observation
+        
+        return observation, reward, terminated, truncated, info
+
+    def __experience_initial(self):
+        for _ in range(20_000):
+            self.__experience_replay(explore_only=True)
+        self.tensorized_obs = self.__tensorize_observation(self.env.reset())
+
+    def __copy_q_to_target(self):
```
Comment (magsteen) [inline agent.py]: as the q network is called policy network this should be called __copy_policy_to_target

File: snake_state.py [hunk 14]
```diff
@@ -205,6 +204,9 @@ def update(self):
         
         ate = self._snake_step()
         return ate
+    
+    def kill(self):
```
Comment (ngorke) [inline snake_state.py]: rename to __kill

File: agent.py (hunk 2 above)
Comment (chrisryda) [reply agent.py]: I fixed now, please check up

File: agent.py (hunk 3 above)
Comment (chrisryda) [reply agent.py:288]: I fixed now, please check up

File: agent.py (hunk 6 above)
Comment (chrisryda) [reply agent.py]: Added suggestion now!

**Review by magsteen — CHANGES_REQUESTED** [review]

File: agent.py (hunk 7 above)
Comment (chrisryda) [reply agent.py]: Added suggestion now!

File: agent.py (hunk 9 above)
Comment (chrisryda) [reply agent.py]: Fixed!

File: snake_state.py (hunk 14 above)
Comment (chrisryda) [reply snake_state.py]: fixed!

File: agent.py (hunk 1 above)
Comment (chrisryda) [reply agent.py]: Fixed!

File: agent.py (hunk 10 above)
Comment (chrisryda) [reply agent.py]: Ignoring this by mutual agreement

File: agent.py (hunk 12 above)
Comment (chrisryda) [reply agent.py]: Fixed, good spot!

File: agent.py (hunk 13 above)
Comment (chrisryda) [reply agent.py]: Another good spot - fixed!

**Review by chrisryda — APPROVED** [review]

**Review by magsteen — APPROVED** [review]

**Pull request merged**